### PR TITLE
fix #279967: image resize not correctly honoring aspect ratio

### PR DESCRIPTION
**Pull request by Howard-C**
Resolves: https://musescore.org/node/279967.

A new member variable `_aspectRatio` of `InspectorImage` is added to remember the image's original ratio. It is also updated when "Lock aspect ratio" is checked in `lockAspectRatioClicked()`. The new two slots `widthChanged()` and `heightChanged()` help changing the value other than the one changed by user.

### CODE REVIEW

File: mscore/inspector/inspectorImage.cpp [hunk 1]
```diff
@@ -56,10 +62,42 @@ InspectorImage::InspectorImage(QWidget* parent)
 
 void InspectorImage::valueChanged(int idx)
       {
-      InspectorBase::valueChanged(idx);
+      InspectorElementBase::valueChanged(idx);
```
Comment (Spire42) [inline mscore/inspector/inspectorImage.cpp]: Unrelated bug fix?

Comment (Howard-C) [reply mscore/inspector/inspectorImage.cpp]: Well, I saw that and thought this needed to be changed, so I changed it ;-)
Do you mean I should separate it from this commit? I don't see the need actually.

File: mscore/inspector/inspectorImage.h [hunk 2]
```diff
@@ -27,11 +27,15 @@ class InspectorImage : public InspectorElementBase {
       Q_OBJECT
 
       Ui::InspectorImage b;
+      qreal _ratio; // used for widthChanged/heightChanged
```
Comment (Spire42) [inline mscore/inspector/inspectorImage.h]: I suggest that this be renamed to `_aspectRatio` for clarity.

Comment (Howard-C) [reply mscore/inspector/inspectorImage.h]: OK

File: mscore/inspector/inspectorImage.cpp [hunk 3]
```diff
@@ -56,10 +62,42 @@ InspectorImage::InspectorImage(QWidget* parent)
...
       setElement();     // DEBUG
       }
 
+//---------------------------------------------------------
+//   lockAspectRatioClicked
+//---------------------------------------------------------
+
+void InspectorImage::lockAspectRatioClicked(bool clicked)
```
Comment (Spire42) [inline mscore/inspector/inspectorImage.cpp]: Should be:

`void InspectorImage::lockAspectRatioClicked(bool checked)`

Also, I suggest that you throw away the aspect ratio (i.e., reset it to a known ”invalid” value such as `0.0`) when `checked` is `false`.

Comment (Howard-C) [reply mscore/inspector/inspectorImage.cpp]: It should be `clicked` because that's what Qt uses.

And is there any real-world usage of throwing away the aspect ratio when it is `false`? If not, then I don't the need of making the code more complex.

Comment (Spire42) [reply mscore/inspector/inspectorImage.cpp]: > It should be `clicked` because that's what Qt uses.

No, it should be `checked` because that's [what Qt uses](https://doc.qt.io/qt-5/qabstractbutton.html#clicked) and also because it's what makes sense. Think about it: When this function is called, it's always because the checkbox has been clicked (or otherwise manipulated by the user), but the checkbox is not always checked. That's what the parameter is there to tell you.

> And is there any real-world usage of throwing away the aspect ratio when it is `false`?

It's an additional sanity check, and it will also be useful when saving to the score file (i.e., a value of `0` means don't save it).

File: mscore/inspector/inspectorImage.cpp (hunk 1 above)
Comment (Spire42) [reply mscore/inspector/inspectorImage.cpp]: One of the boxes that you're supposed to check when creating a PR affirms that you didn't make any unnecessary changes. I try my best to follow that, and I assumed that everyone else did too.

Comment (Howard-C) [reply mscore/inspector/inspectorImage.cpp]: This is a _necessary_ change to ensure the function of the correct parent class being called.

File: mscore/inspector/inspectorImage.cpp (hunk 3 above)
Comment (Howard-C) [reply mscore/inspector/inspectorImage.cpp]: I'm convinced of the first matter.

For the second matter, I think "additional sanity check" should only take place if it can prevent latent failures or crashes. And as I said, this fix has nothing to do with saving to score file. If a variable is intended to be saved to score, it should be defined in `libmscore` rather than `mscore`.

File: mscore/inspector/inspectorImage.cpp (hunk 1 above)
Comment (Spire42) [reply mscore/inspector/inspectorImage.cpp]: I take “necessary” to mean in the context of what's being fixed in the PR. Unless I misread the code, the parent class isn't actually doing anything in this case, so this fix, while good, is not necessary for this PR.

I don't know how much of a stickler we're expected to be, but I personally err on the side of caution.

Comment (Howard-C) [reply mscore/inspector/inspectorImage.cpp]: I've always been doing these changes which, although irrelevant to the issue being resolved, are good and cannot possibly have any adverse impact. It seems a bit much to create another commit just for this one-line change, and if we don't change this now, it may lead to a bug later on when the functionality of `InspectorElementBase::valueChanged()` is expanded.

Comment (Spire42) [reply mscore/inspector/inspectorImage.cpp]: Fair enough. Just thought I'd ask.

**Comment by anatoly-os** [comment]
@Howard-C I would leave inspector related changes to the new inspector. Moreover, this issue has probably been addressed there. What do you think?

**Comment by Howard-C** [comment]
Not fixing it for 3.5?

**Comment by Howard-C** [comment]
I'd suggest this go to 3.6
